crates/red_knot_python_semantic/src/semantic_index/builder/except_handlers.rs

```diff
@@ -0,0 +1,128 @@
+use std::cell::{RefCell, RefMut};
+
+use crate::semantic_index::use_def::FlowSnapshot;
+
+use super::SemanticIndexBuilder;
+
+/// An abstraction over the fact that each scope should have its own [`TryNodeContextStack`]
+#[derive(Debug, Default)]
+pub(super) struct TryNodeContextStackManager(Vec<TryNodeContextStack>);
+
+impl TryNodeContextStackManager {
+    /// Push a new [`TryNodeContextStack`] onto the stack of stacks.
+    ///
+    /// Each [`TryNodeContextStack`] is only valid for a single scope
+    pub(super) fn enter_nested_scope(&mut self) {
+        self.0.push(TryNodeContextStack::default());
+    }
+
+    /// Retrieve the [`TryNodeContextStack`] that is relevant for the current scope.
+    pub(super) fn current_try_context_stack(&self) -> &TryNodeContextStack {
+        self.0
+            .last()
+            .expect("There should always be at least one `TryBlockContexts` on the stack")
+    }
+
+    /// Pop a new [`TryNodeContextStack`] off the stack of stacks.
+    ///
+    /// Each [`TryNodeContextStack`] is only valid for a single scope
+    pub(super) fn exit_scope(&mut self) {
+        let popped_context = self.0.pop();
+        assert!(
+            popped_context.is_some(),
+            "pop_context() should never be called on an empty stack \
+(this indicates an unbalanced `push_context()`/`pop_context()` pair of calls)"
+        );
+    }
+}
+
+/// The contexts of nested `try`/`except` blocks for a single scope
+#[derive(Debug, Default)]
+pub(super) struct TryNodeContextStack(RefCell<Vec<TryNodeContext>>);
```

Comment:
Removing the `RefCell` and instead rely on compile time borrow checking doesn't seem to be difficult. I quickly tried it out locally and the only issue you run into is with `record_definition`. My recommendation there is to instead expose an `iter_mut` method on `TryNodeStack` and inline the logic into `builder.rs`. 

Removing the `RefCell` should also remove the need for `TryNodeContextStackRefMut` because you can just return a `&TryNodeContextStack` or `&mut TryNodeContextStack` based on the mutability that is needed.

Reply:
I had to bring back the `RefCell` in https://github.com/astral-sh/ruff/pull/13729/commits/24d139891efc6030ef0f3ec2eebd1999c2e6e831. `record_definition()` is indeed the only place where normal borrowing rules are difficult to work with here, but I still don't see how to avoid using a `RefCell` internally :( I tried your suggestion of exposing an `iter_mut` method and inlining the logic into `builder.rs`, but I still couldn't make the borrow checker happy.

I have at least managed to remove `TryNodeContextStackRefMut`, however!

Reply:
Yeah, the `self.flow_snapshot` makes this a bit awkward. There are two possibilities:

1. Pass the snapshot to `record_definition`

```patch
Subject: [PATCH] Update snapshot

The old snapshot was incorrect because the format range excluded the first (all empty) line. Now the first line gets formatted also.
---
Index: crates/red_knot_python_semantic/src/semantic_index/builder.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/crates/red_knot_python_semantic/src/semantic_index/builder.rs b/crates/red_knot_python_semantic/src/semantic_index/builder.rs
--- a/crates/red_knot_python_semantic/src/semantic_index/builder.rs	(revision 24d139891efc6030ef0f3ec2eebd1999c2e6e831)
+++ b/crates/red_knot_python_semantic/src/semantic_index/builder.rs	(date 1729073072841)
@@ -103,9 +103,9 @@
             .expect("Always to have a root scope")
     }
 
-    fn try_node_context_stack(&self) -> &TryNodeContextStack {
+    fn try_node_context_stack(&mut self) -> &mut TryNodeContextStack {
         self.try_node_context_stack_manager
-            .current_try_context_stack()
+            .current_try_context_stack_mut()
     }
 
     fn push_scope(&mut self, node: NodeWithScopeRef) {
@@ -239,7 +239,9 @@
             DefinitionCategory::Declaration => use_def.record_declaration(symbol, definition),
             DefinitionCategory::Binding => use_def.record_binding(symbol, definition),
         }
-        self.try_node_context_stack().record_definition(self);
+
+        let snapshot = self.flow_snapshot();
+        self.try_node_context_stack().record_definition(snapshot);
 
         definition
     }
Index: crates/red_knot_python_semantic/src/semantic_index/builder/except_handlers.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/crates/red_knot_python_semantic/src/semantic_index/builder/except_handlers.rs b/crates/red_knot_python_semantic/src/semantic_index/builder/except_handlers.rs
--- a/crates/red_knot_python_semantic/src/semantic_index/builder/except_handlers.rs	(revision 24d139891efc6030ef0f3ec2eebd1999c2e6e831)
+++ b/crates/red_knot_python_semantic/src/semantic_index/builder/except_handlers.rs	(date 1729073033138)
@@ -23,6 +23,12 @@
             .expect("There should always be at least one `TryBlockContexts` on the stack")
     }
 
+    pub(super) fn current_try_context_stack_mut(&mut self) -> &mut TryNodeContextStack {
+        self.0
+            .last_mut()
+            .expect("There should always be at least one `TryBlockContexts` on the stack")
+    }
+
     /// Pop a new [`TryNodeContextStack`] off the stack of stacks.
     ///
     /// Each [`TryNodeContextStack`] is only valid for a single scope
@@ -38,33 +44,36 @@
 
 /// The contexts of nested `try`/`except` blocks for a single scope
 #[derive(Debug, Default)]
-pub(super) struct TryNodeContextStack(RefCell<Vec<TryNodeContext>>);
+pub(super) struct TryNodeContextStack(Vec<TryNodeContext>);
 
 impl TryNodeContextStack {
     /// Push a new [`TryNodeContext`] for recording intermediate states
     /// while visiting a [`ruff_python_ast::StmtTry`] node that has a `finally` branch.
-    pub(super) fn push_context(&self) {
-        self.0.borrow_mut().push(TryNodeContext::default());
+    pub(super) fn push_context(&mut self) {
+        self.0.push(TryNodeContext::default());
     }
 
     /// Pop a [`TryNodeContext`] off the stack.
-    pub(super) fn pop_context(&self) -> Vec<FlowSnapshot> {
+    pub(super) fn pop_context(&mut self) -> Vec<FlowSnapshot> {
         let TryNodeContext {
             try_suite_snapshots,
         } = self
             .0
-            .borrow_mut()
             .pop()
             .expect("Cannot pop a `try` block off an empty `TryBlockContexts` stack");
         try_suite_snapshots
     }
 
     /// For each `try` block on the stack, push the snapshot onto the `try` block
-    pub(super) fn record_definition(&self, builder: &SemanticIndexBuilder) {
-        for context in self.0.borrow_mut().iter_mut() {
-            context.record_definition(builder.flow_snapshot());
+    pub(super) fn record_definition(&mut self, snapshot: FlowSnapshot) {
+        for context in self.0.iter_mut() {
+            context.record_definition(snapshot.clone());
         }
     }
+
+    pub(super) fn iter_mut(&mut self) -> impl Iterator<Item = &mut TryNodeContext> {
+        self.0.iter_mut()
+    }
 }
 
 /// Context for tracking definitions over the course of a single
```

2. do the "move" trick

```rust
let use_def = self.current_use_def_map_mut();
        match category {
            DefinitionCategory::DeclarationAndBinding => {
                use_def.record_declaration_and_binding(symbol, definition);
            }
            DefinitionCategory::Declaration => use_def.record_declaration(symbol, definition),
            DefinitionCategory::Binding => use_def.record_binding(symbol, definition),
        }

        let mut try_stack_manager = std::mem::take(&mut self.try_node_context_stack_manager);
        for context in try_stack_manager.current_try_context_stack_mut().iter_mut() {
            context.record_definition(self.flow_snapshot());
        }
        self.try_node_context_stack_manager = try_stack_manager;

        definition
```

Reply:
Idea (1) is what I did last night and it caused a massive perf regression, because it means you end up taking a snapshot after _every_ definition, even if you're not visiting a node inside an exception handler!

I'll try out idea (2) and see if it works